File: Algorithms/0589_N-ary_Tree_Preorder_Traversal/Python/N-ary_Tree_Preorder_Traversal_Solution_1.py

# Space: O(n)
# Time: O(n)

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root):
        if root is None: return []
        if root.children is None: return root.val
        queue = [root]
        res = []
        while queue:
            cur_node = queue.pop(0)
            res.append(cur_node.val)
            if cur_node.children:
                queue = cur_node.children + queue

        return res

File: Algorithms/0589_N-ary_Tree_Preorder_Traversal/Java/N-AryTreePreorderTraversalSolution1.java

// Space: O(n)
// Time: O(n)

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {

    public List<Integer> preorder(Node root) {
        if (root == null) return new ArrayList<>();
        if (root.children == null) return new ArrayList<Integer>(Arrays.asList(root.val));

        List<Integer> res = new ArrayList<>();
        List<Node> queue = new ArrayList<>();
        queue.add(root);

        while (queue.size() > 0) {
            Node cur = queue.remove(0);
            res.add(cur.val);
            if (cur.children != null) {
                cur.children.addAll(queue);
                queue = cur.children;
            }
        }
        return res;
    }
}




